diff --git a/easyrsa3/easyrsa b/easyrsa3/easyrsa
index 3e5dcc1..626999a 100755
--- a/easyrsa3/easyrsa
+++ b/easyrsa3/easyrsa
@@ -1147,13 +1147,8 @@ install_data_to_pki() {
 
 	# Find and optionally copy data-files, in specific order
 	for area in \
-		"$EASYRSA_PKI" \
-		"$EASYRSA" \
 		"$PWD" \
-		"${0%/*}" \
-		'/usr/local/share/easy-rsa' \
-		'/usr/share/easy-rsa' \
-		'/etc/easy-rsa' \
+		'@out@/share/easyrsa/'
 		# EOL
 	do
 		if [ "$context" = x509-types-only ]; then
@@ -4470,7 +4465,7 @@ Sourcing the vars file and building certificates will probably fail ..'
 
 	# Set defaults, preferring existing env-vars if present
 	set_var EASYRSA					"$PWD"
-	set_var EASYRSA_OPENSSL			openssl
+	set_var EASYRSA_OPENSSL			@openssl@
 	set_var EASYRSA_PKI				"$EASYRSA/pki"
 	set_var EASYRSA_DN				cn_only
 	set_var EASYRSA_REQ_COUNTRY		"US"
@@ -4505,8 +4500,8 @@ Sourcing the vars file and building certificates will probably fail ..'
 	set_var EASYRSA_REQ_CN			ChangeMe
 	set_var EASYRSA_DIGEST			sha256
 
-	set_var EASYRSA_SSL_CONF		"$EASYRSA_PKI/openssl-easyrsa.cnf"
-	set_var EASYRSA_SAFE_CONF		"$EASYRSA_PKI/safessl-easyrsa.cnf"
+	set_var EASYRSA_SSL_CONF        "@out@/share/easyrsa/openssl-easyrsa.cnf"
+	set_var EASYRSA_SAFE_CONF       "@out@/share/easyrsa/safessl-easyrsa.cnf"
 
 	set_var EASYRSA_KDC_REALM		"CHANGEME.EXAMPLE.COM"
 
